examples/karpenter-fargate/main.tf

```diff
@@ -0,0 +1,338 @@
+  # Security groups used in this module created by the upstream modules terraform-aws-eks (https://github.com/terraform-aws-modules/terraform-aws-eks).
+  #   Upstream module implemented Security groups based on the best practices doc https://docs.aws.amazon.com/eks/latest/userguide/sec-group-reqs.html.
+  #   So, by default the security groups are restrictive. Users needs to enable rules for specific ports required for App requirement or Add-ons
+  #   See the notes below for each rule used in these examples
+  #----------------------------------------------------------------------------------------------------------#
+  node_security_group_additional_rules = {
+    # Extend node-to-node security group rules. Recommended and required for the Add-ons
+    ingress_self_all = {
+      description = "Node to node all ports/protocols"
+      protocol    = "-1"
+      from_port   = 0
+      to_port     = 0
+      type        = "ingress"
+      self        = true
+    }
+    # Recommended outbound traffic for Node groups
+    egress_all = {
+      description      = "Node all egress"
+      protocol         = "-1"
+      from_port        = 0
+      to_port          = 0
+      type             = "egress"
+      cidr_blocks      = ["0.0.0.0/0"]
+      ipv6_cidr_blocks = ["::/0"]
+    }
+
+    # Allows Control Plane Nodes to talk to Worker nodes on Karpenter ports.
+    # This can be extended further to specific port based on the requirement for others Add-on e.g., metrics-server 4443, spark-operator 8080, etc.
+    # Change this according to your security requirements if needed
+    ingress_nodes_karpenter_port = {
+      description                   = "Cluster API to Nodegroup for Karpenter"
+      protocol                      = "tcp"
+      from_port                     = 8443
+      to_port                       = 8443
+      type                          = "ingress"
+      source_cluster_security_group = true
+    }
+  }
+
+  # Add karpenter.sh/discovery tag so that we can use this as securityGroupSelector in karpenter provisioner
+  node_security_group_tags = {
+    "karpenter.sh/discovery/${local.name}" = local.name
+  }
+
+  # Add Karpenter IAM role to the aws-auth config map to allow the controller to register the ndoes to the clsuter
+  map_roles = [
+    {
+      rolearn  = aws_iam_role.karpenter.arn
+      username = "system:node:{{EC2PrivateDNSName}}"
+      groups = [
+        "system:bootstrappers",
+        "system:nodes"
+      ]
+    }
+  ]
+
+  # EKS FARGATE PROFILES
+  # We recommend to have Fargate profiles to place your critical workloads and add-ons
+  # Then rely on Karpenter to scale your workloads
+  # We filter the kube-system pods with labels since not all add-ons can run on Fargate (e.g. aws-node-termination-handler)
+  fargate_profiles = {
+    # Providing compute for the kube-system namespace where addons that can run on Fargate reside
+    coredns = {
```

Comment:
since karpenter does not depend on coredns, it seems best to have it running on karpenter nodes than the lead time to create and maintain more fargate profiles, roles, etc

Reply:
For this example, I decided to put the coredns and also the aws-load-balancer-controller additions on Fargate. Not because Karpenter depends on them, but because I think Fargate might be a better choice to run these add-ons since they don't need specific hardware support or a particular instance type. For these reasons, my guiding principle was to put addons on Fargate unless they were not compatible with it, like NTH.

Reply:
i just found that creating and destroying multiple fargate profiles to take far too long, since they are done in serial rather than parallel, and karpenter nodes are faster than that.

Reply:
I created this example while thinking about how AWS customers could use it in the real world. I don't think the time it takes to create and destroy the resources provided by this example should be taken into consideration.

Reply:
@FernandoMiguel have you tried [profile wildcards](https://docs.aws.amazon.com/eks/latest/userguide/fargate-profile.html#fargate-profile-wildcards) to work the problem?

Reply:
If we remove NTH or deploy in another namespace we will be able to use fargate profile wildcards. I did that since the NTH namespace is hardcoded to  `kube-system` in https://github.com/aws-ia/terraform-aws-eks-blueprints/blob/main/modules/kubernetes-addons/aws-node-termination-handler/locals.tf#L2 and I didn't want to introduce that change in this PR. Since NTH will not be required with Karpenter v.0.19.0 I think we should just remove it from that PR like I proposed in https://github.com/aws-ia/terraform-aws-eks-blueprints/pull/1187#discussion_r1025505127.

Reply:
Since Karpenter 0.20.0 is now available in the blueprints I removed NTH and this example now uses only two Fargate profiles.